Mock server support for rolling restart

### Describe the bug

Attempt to execute rolling restart on the Mock server,  failing with error: `Message: Unsupported Media Type.`

### Fabric8 Kubernetes Client version

6.13.4

### Steps to reproduce

* Configure Mock server in crud mode `@EnableKubernetesMockClient(crud = true)`
* Create resource, for example Deployment
* Execute rolling restart:
```
kubernetesClient.apps()
      .deployments()
      .inNamespace("ns")
      .withName("app01r")
      .rolling().restart();
```




### Expected behavior

to work, ;)

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.25.3@latest

### Environment

macOS

### Fabric8 Kubernetes Client Logs

```shell
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://localhost:57979/apis/apps/v1/namespaces/infra/deployments/mike-vd-server. Message: Unsupported Media Type. Received status: Status(apiVersion=v1, code=415, details=StatusDetails(causes=[], group=null, kind=Unknown, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Unsupported Media Type, metadata=null, reason=Invalid, status=Failure, additionalProperties={}).

	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:419)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:268)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:44)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.RollingUpdater.applyPatch(RollingUpdater.java:173)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.RollingUpdater.restart(RollingUpdater.java:185)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.DeploymentOperationsImpl.restart(DeploymentOperationsImpl.java:102)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.DeploymentOperationsImpl.restart(DeploymentOperationsImpl.java:52)
	at com.opentable.instanceno.machine.KubernetesClientTest.testGetDeployment(KubernetesClientTest.java:121)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://localhost:57979/apis/apps/v1/namespaces/infra/deployments/mike-vd-server. Message: Unsupported Media Type. Received status: Status(apiVersion=v1, code=415, details=StatusDetails(causes=[], group=null, kind=Unknown, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Unsupported Media Type, metadata=null, reason=Invalid, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:660)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:640)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:589)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:141)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:51)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:136)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```


### Additional context

It seems like rolling restarts using `application/strategic-merge-patch+json` https://github.com/fabric8io/kubernetes-client/pull/2272/files, but Mock server not supporting strategic merges.

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/RollingUpdater.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/RollingUpdater.java
@@ -205,8 +205,9 @@ public abstract class RollingUpdater<T extends HasMetadata, L> {
     HashMap<String, Object> value = new HashMap<>();
     patch.put("op", "replace");
     patch.put("path", "/spec/template/metadata/annotations");
-    value.put("kubectl.kubernetes.io/restartedAt", new Date().toInstant().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
-    patch.put("value",  value);
+    value.put("kubectl.kubernetes.io/restartedAt",
+        new Date().toInstant().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
+    patch.put("value", value);
     return Collections.singletonList(patch);
   }
 

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/RollingUpdater.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/RollingUpdater.java
@@ -45,10 +45,10 @@ import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.time.ZoneOffset;
 import java.time.format.DateTimeFormatter;
+import java.util.Collections;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Future;
@@ -169,8 +169,8 @@ public abstract class RollingUpdater<T extends HasMetadata, L> {
     }
   }
 
-  private static <T> T applyPatch(Resource<T> resource, Map<String, Object> map, KubernetesSerialization serialization) {
-    return resource.patch(PatchContext.of(PatchType.STRATEGIC_MERGE), serialization.asJson(map));
+  private static <T> T applyPatch(Resource<T> resource, List<Object> list, KubernetesSerialization serialization) {
+    return resource.patch(PatchContext.of(PatchType.JSON), serialization.asJson(list));
   }
 
   public static <T extends HasMetadata> T resume(RollableScalableResourceOperation<T, ?, ?> resource) {
@@ -185,35 +185,29 @@ public abstract class RollingUpdater<T extends HasMetadata, L> {
     return applyPatch(resource, RollingUpdater.requestPayLoadForRolloutRestart(), resource.getKubernetesSerialization());
   }
 
-  public static Map<String, Object> requestPayLoadForRolloutPause() {
-    Map<String, Object> jsonPatchPayload = new HashMap<>();
-    Map<String, Object> spec = new HashMap<>();
-    spec.put("paused", true);
-    jsonPatchPayload.put("spec", spec);
-    return jsonPatchPayload;
+  public static List<Object> requestPayLoadForRolloutPause() {
+    HashMap<String, Object> patch = new HashMap<>();
+    patch.put("op", "add");
+    patch.put("path", "/spec/paused");
+    patch.put("value", true);
+    return Collections.singletonList(patch);
   }
 
-  public static Map<String, Object> requestPayLoadForRolloutResume() {
-    Map<String, Object> jsonPatchPayload = new HashMap<>();
-    Map<String, Object> spec = new HashMap<>();
-    spec.put("paused", null);
-    jsonPatchPayload.put("spec", spec);
-    return jsonPatchPayload;
+  public static List<Object> requestPayLoadForRolloutResume() {
+    HashMap<String, Object> patch = new HashMap<>();
+    patch.put("op", "remove");
+    patch.put("path", "/spec/paused");
+    return Collections.singletonList(patch);
   }
 
-  public static Map<String, Object> requestPayLoadForRolloutRestart() {
-    Map<String, Object> jsonPatchPayload = new HashMap<>();
-    Map<String, String> annotations = new HashMap<>();
-    annotations.put("kubectl.kubernetes.io/restartedAt",
-        new Date().toInstant().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
-    Map<String, Object> templateMetadata = new HashMap<>();
-    templateMetadata.put("annotations", annotations);
-    Map<String, Object> template = new HashMap<>();
-    template.put("metadata", templateMetadata);
-    Map<String, Object> deploymentSpec = new HashMap<>();
-    deploymentSpec.put("template", template);
-    jsonPatchPayload.put("spec", deploymentSpec);
-    return jsonPatchPayload;
+  public static List<Object> requestPayLoadForRolloutRestart() {
+    HashMap<String, Object> patch = new HashMap<>();
+    HashMap<String, Object> value = new HashMap<>();
+    patch.put("op", "replace");
+    patch.put("path", "/spec/template/metadata/annotations");
+    value.put("kubectl.kubernetes.io/restartedAt", new Date().toInstant().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
+    patch.put("value",  value);
+    return Collections.singletonList(patch);
   }
 
   /**

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DeploymentCrudTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DeploymentCrudTest.java
@@ -132,22 +132,22 @@ class DeploymentCrudTest {
   void testPause() {
     // Given
     Deployment deployment1 = new DeploymentBuilder().withNewMetadata()
-      .withName("d1")
-      .withNamespace("ns1")
-      .addToLabels("testKey", "testValue")
-      .endMetadata()
-      .withNewSpec()
-      .endSpec()
-      .build();
+        .withName("d1")
+        .withNamespace("ns1")
+        .addToLabels("testKey", "testValue")
+        .endMetadata()
+        .withNewSpec()
+        .endSpec()
+        .build();
     client.apps().deployments().inNamespace("ns1").create(deployment1);
 
     // When
     client.apps()
-      .deployments()
-      .inNamespace("ns1")
-      .withName("d1")
-      .rolling()
-      .pause();
+        .deployments()
+        .inNamespace("ns1")
+        .withName("d1")
+        .rolling()
+        .pause();
     Deployment updatedDeployment = client.apps().deployments().inNamespace("ns1").withName("d1").get();
 
     // Then
@@ -160,23 +160,23 @@ class DeploymentCrudTest {
   void testRolloutResume() throws InterruptedException {
     // Given
     Deployment deployment1 = new DeploymentBuilder().withNewMetadata()
-      .withName("d1")
-      .withNamespace("ns1")
-      .addToLabels("testKey", "testValue")
-      .endMetadata()
-      .withNewSpec()
-      .withPaused(true)
-      .endSpec()
-      .build();
+        .withName("d1")
+        .withNamespace("ns1")
+        .addToLabels("testKey", "testValue")
+        .endMetadata()
+        .withNewSpec()
+        .withPaused(true)
+        .endSpec()
+        .build();
     client.apps().deployments().inNamespace("ns1").create(deployment1);
 
     // When
     client.apps()
-      .deployments()
-      .inNamespace("ns1")
-      .withName("d1")
-      .rolling()
-      .resume();
+        .deployments()
+        .inNamespace("ns1")
+        .withName("d1")
+        .rolling()
+        .resume();
     Deployment updatedDeployment = client.apps().deployments().inNamespace("ns1").withName("d1").get();
 
     // Then
@@ -189,31 +189,32 @@ class DeploymentCrudTest {
   void testRolloutRestart() throws InterruptedException {
     // Given
     Deployment deployment1 = new DeploymentBuilder().withNewMetadata()
-      .withName("d1")
-      .withNamespace("ns1")
-      .addToLabels("testKey", "testValue")
-      .endMetadata()
-      .withNewSpec()
-      .withNewTemplate()
-      .withNewMetadata()
-      .addToAnnotations("", "")
-      .endMetadata()
-      .endTemplate()
-      .endSpec()
-      .build();
+        .withName("d1")
+        .withNamespace("ns1")
+        .addToLabels("testKey", "testValue")
+        .endMetadata()
+        .withNewSpec()
+        .withNewTemplate()
+        .withNewMetadata()
+        .addToAnnotations("", "")
+        .endMetadata()
+        .endTemplate()
+        .endSpec()
+        .build();
     client.apps().deployments().inNamespace("ns1").create(deployment1);
 
     // When
     client.apps()
-      .deployments()
-      .inNamespace("ns1")
-      .withName("d1")
-      .rolling()
-      .restart();
+        .deployments()
+        .inNamespace("ns1")
+        .withName("d1")
+        .rolling()
+        .restart();
     Deployment updatedDeployment = client.apps().deployments().inNamespace("ns1").withName("d1").get();
 
     // Then
     assertNotNull(updatedDeployment);
-    assertNotNull(updatedDeployment.getSpec().getTemplate().getMetadata().getAnnotations().get("kubectl.kubernetes.io/restartedAt"));
+    assertNotNull(
+        updatedDeployment.getSpec().getTemplate().getMetadata().getAnnotations().get("kubectl.kubernetes.io/restartedAt"));
   }
 }

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DeploymentCrudTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DeploymentCrudTest.java
@@ -30,6 +30,7 @@ import java.util.Collections;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @EnableKubernetesMockClient(crud = true)
@@ -124,5 +125,95 @@ class DeploymentCrudTest {
     assertNotNull(replacedDeployment);
     assertFalse(replacedDeployment.getMetadata().getLabels().isEmpty());
     assertEquals("one", replacedDeployment.getMetadata().getLabels().get("testKey"));
+  }
+
+  @Test
+  @DisplayName("Should pause resource")
+  void testPause() {
+    // Given
+    Deployment deployment1 = new DeploymentBuilder().withNewMetadata()
+      .withName("d1")
+      .withNamespace("ns1")
+      .addToLabels("testKey", "testValue")
+      .endMetadata()
+      .withNewSpec()
+      .endSpec()
+      .build();
+    client.apps().deployments().inNamespace("ns1").create(deployment1);
+
+    // When
+    client.apps()
+      .deployments()
+      .inNamespace("ns1")
+      .withName("d1")
+      .rolling()
+      .pause();
+    Deployment updatedDeployment = client.apps().deployments().inNamespace("ns1").withName("d1").get();
+
+    // Then
+    assertNotNull(updatedDeployment);
+    assertEquals(true, updatedDeployment.getSpec().getPaused());
+  }
+
+  @Test
+  @DisplayName("Should resume rollout")
+  void testRolloutResume() throws InterruptedException {
+    // Given
+    Deployment deployment1 = new DeploymentBuilder().withNewMetadata()
+      .withName("d1")
+      .withNamespace("ns1")
+      .addToLabels("testKey", "testValue")
+      .endMetadata()
+      .withNewSpec()
+      .withPaused(true)
+      .endSpec()
+      .build();
+    client.apps().deployments().inNamespace("ns1").create(deployment1);
+
+    // When
+    client.apps()
+      .deployments()
+      .inNamespace("ns1")
+      .withName("d1")
+      .rolling()
+      .resume();
+    Deployment updatedDeployment = client.apps().deployments().inNamespace("ns1").withName("d1").get();
+
+    // Then
+    assertNotNull(updatedDeployment);
+    assertNull(updatedDeployment.getSpec().getPaused());
+  }
+
+  @Test
+  @DisplayName("Should restart rollout")
+  void testRolloutRestart() throws InterruptedException {
+    // Given
+    Deployment deployment1 = new DeploymentBuilder().withNewMetadata()
+      .withName("d1")
+      .withNamespace("ns1")
+      .addToLabels("testKey", "testValue")
+      .endMetadata()
+      .withNewSpec()
+      .withNewTemplate()
+      .withNewMetadata()
+      .addToAnnotations("", "")
+      .endMetadata()
+      .endTemplate()
+      .endSpec()
+      .build();
+    client.apps().deployments().inNamespace("ns1").create(deployment1);
+
+    // When
+    client.apps()
+      .deployments()
+      .inNamespace("ns1")
+      .withName("d1")
+      .rolling()
+      .restart();
+    Deployment updatedDeployment = client.apps().deployments().inNamespace("ns1").withName("d1").get();
+
+    // Then
+    assertNotNull(updatedDeployment);
+    assertNotNull(updatedDeployment.getSpec().getTemplate().getMetadata().getAnnotations().get("kubectl.kubernetes.io/restartedAt"));
   }
 }

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DeploymentTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DeploymentTest.java
@@ -622,7 +622,7 @@ class DeploymentTest {
     // Then
     RecordedRequest recordedRequest = server.getLastRequest();
     assertEquals("PATCH", recordedRequest.getMethod());
-    assertEquals("{\"spec\":{\"paused\":true}}", recordedRequest.getBody().readUtf8());
+    assertEquals("[{\"op\":\"add\",\"path\":\"/spec/paused\",\"value\":true}]", recordedRequest.getBody().readUtf8());
   }
 
   @Test
@@ -652,7 +652,7 @@ class DeploymentTest {
     RecordedRequest recordedRequest = server.getLastRequest();
     assertNotNull(deployment);
     assertEquals("PATCH", recordedRequest.getMethod());
-    assertEquals("{\"spec\":{\"paused\":null}}", recordedRequest.getBody().readUtf8());
+    assertEquals("[{\"op\":\"remove\",\"path\":\"/spec/paused\"}]", recordedRequest.getBody().readUtf8());
   }
 
   @Test

